builder/format.go

```diff
@@ -0,0 +1,129 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package builder
+
+import (
+	"io/ioutil"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/clearlinux/mixer-tools/helpers"
+	"github.com/pkg/errors"
+)
+
+// UpdateFormatVersion updates the builder.conf file with a new format version
+func UpdateFormatVersion(b *Builder, version string) error {
+	b.Config.Swupd.Format = version
+
+	newver := "${1}" + b.Config.Swupd.Format
+	var re = regexp.MustCompile(`(FORMAT=)[0-9]*`)
+
+	builderData, err := ioutil.ReadFile(b.BuildConf)
+	if err != nil {
+		return errors.Wrap(err, "Failed to read builder.conf")
+	}
+
+	builderEdit := re.ReplaceAllString(string(builderData), newver)
+
+	var filename string
+	if UseNewConfig {
+		filename, err = GetConfigPath("")
+		if err != nil {
+			return err
+		}
+
+		return b.Config.SaveConfig(filename)
+	}
+
+	builderOut := []byte(builderEdit)
+	if err = ioutil.WriteFile(b.BuildConf, builderOut, 0644); err != nil {
+		return errors.Wrap(err, "Failed to write new builder.conf")
+	}
+
+	return nil
+}
+
+// CopyFullGroupsINI copies the initial ini file which has ALL bundle definitions
+func CopyFullGroupsINI(b *Builder) error {
+	return helpers.CopyFile(filepath.Join(b.Config.Builder.ServerStateDir, "full_groups.ini"), filepath.Join(b.Config.Builder.ServerStateDir, "groups.ini"))
+}
+
+// CopyTrimmedGroupsINI copies the new ini made with deleted bundles removed
+func CopyTrimmedGroupsINI(b *Builder) error {
+	return helpers.CopyFile(filepath.Join(b.Config.Builder.ServerStateDir, "trimmed_groups.ini"), filepath.Join(b.Config.Builder.ServerStateDir, "groups.ini"))
+}
+
+// RevertFullGroupsINI copies back the full ini to the manifest creator accounts for deleted bundles
+func RevertFullGroupsINI(b *Builder) error {
+	return helpers.CopyFile(filepath.Join(b.Config.Builder.ServerStateDir, "groups.ini"), filepath.Join(b.Config.Builder.ServerStateDir, "full_groups.ini"))
+}
+
+// RevertTrimmedGroupsINI copies back the trimmed INI so manifests are not created anymore for deleted bundles in new format
+func RevertTrimmedGroupsINI(b *Builder) error {
+	return helpers.CopyFile(filepath.Join(b.Config.Builder.ServerStateDir, "groups.ini"), filepath.Join(b.Config.Builder.ServerStateDir, "trimmed_groups.ini"))
+}
+
+func getLastBuildVersion(b *Builder) (string, error) {
+	var lastVer []byte
+	var err error
+
+	filename := filepath.Join(b.Config.Builder.ServerStateDir, "image/LAST_VER")
+	// Likely the first build or some weirdness that requires us to get the container image and build it anyway
+	if lastVer, err = ioutil.ReadFile(filename); err != nil {
+		return "", errors.Wrap(err, "Cannot find last built version, must get container image")
```

Comment:
So what does this comment/error message actually mean? Doesn't this case _always_ mean it's the first build? Is it actually possible for a build to succeed without generating this file? And I'm not following what difference running in a container would make.

Reply:
It shouldn't be possible for it to succeed without this file existing. It could mean someone deleted it or broke it too.
However, running a build with this file not existing is not an error case, it will re-create the file and take it as a "build from 0" build and work successfully, which means the container creating the build will succeed in making an update.